shapely/coordinates.py

```diff
@@ -1,30 +1,76 @@
-    """Returns a copy of a geometry array with a function applied to its
-    coordinates.
+import numpy as np
 
-    With the default of ``include_z=False``, all returned geometries will be
-    two-dimensional; the third dimension will be discarded, if present.
-    When specifying ``include_z=True``, the returned geometries preserve
-    the dimensionality of the respective input geometries.
+import shapely
+from shapely import lib
+from shapely.lib import Geometry
+
+__all__ = [
+    "transform",
+    "_transform_multi_vec",
+    "_transform",
+    "_transform2_multi_vec",
+    "_transform2",
+    "count_coordinates",
+    "get_coordinates",
+    "set_coordinates",
+]
+
+from shapely.errors import GeometryTypeError
+
+
+def transform(
+    geometry,
+    transformation,
+    include_z: Union[bool, int] = False,
+    single_vec: bool = True,
+    rebuild: bool = False,
+):
+    """Returns a copy of a geometry array with a function applied to its coordinates.
+
+    With the default of ``rebuild=False``, ``include_z=False``,
+    all returned geometries will be two-dimensional;
+    the third dimension will be discarded, if present.
+    When specifying ``rebuild=False``, ``include_z=True``,
+    the returned geometries preserve the dimensionality of the
+    respective input geometries.
+    When specifying ``rebuild=True`` the output geometries
+    will get rebuilt from the transformation output
 
     Parameters
     ----------
     geometry : Geometry or array_like
     transformation : function
-        A function that transforms a (N, 2) or (N, 3) ndarray of float64 to
+        A function that transforms coordinates, in ``single_vec=True`` mode:
+        the function transforms a (N, 2) or (N, 3) ndarray of float64 to
         another (N, 2) or (N, 3) ndarray of float64.
-    include_z : bool, default False
-        If True, include the third dimension in the coordinates array
-        that is passed to the ``transformation`` function. If a
-        geometry has no third dimension, the z-coordinates passed to the
-        function will be NaN.
+        in ``single_vec=False`` mode: the function transform 2 or 3 N ndarray
+        of float64 to another 2 or 3 ndarray of float64.
+        With the default ``rebuild=False``, the transform function must also
+        return N sized array(s).
+    include_z : bool or int, default False
+        If True | 1, include the third dimension in the coordinates array
+            that is passed to the ``transformation`` function. If a
+            geometry has no third dimension, the z-coordinates passed to the
+            function will be NaN.
+        If False | 0, The third-dimension of the geometries (if exist),
+            would not be included in the input to transform function.
+        If -1 Include the third dimension in the coordinates array that is passed to the
+            ``transformation`` function only if the geoemtries are three-dimensional.
+    single_vec: bool, default True
```

Comment:
In https://github.com/shapely/shapely/pull/1559, I mentioned `interleaved=True/False` as an optional name (it wasn't included in the PR, only mentioned as possible future enhancement). 

That_might_ be a more descriptive name for the keyword (although `single_vec` is more explicit about the _effect_ of using interleaved coordinates or not, i.e. whether the coordinates are passed as a single array or not)